### PR TITLE
feat(): Add support for typeorm@0.3.0 literal repositories #383

### PR DESCRIPTION
**Pull request by matomesc**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [ ] The commit message follows our guidelines: https://github.com/nestjs/nest/blob/master/CONTRIBUTING.md
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Other... Please describe:
```

## What is the current behavior?

Issue Number: https://github.com/nestjs/typeorm/issues/383

## What is the new behavior?
Check if `Repository` is a function to determine if `typeorm@0.3.0` is being used and if so then use the literal repository object to generate repository injection token.

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

## Other information
Tests are passing for current `typeorm@0.2.24` but failing for `typeorm@0.3.0` due to `Repository` no longer being a function. The custom repository also needs to be refactored: https://github.com/nestjs/typeorm/blob/master/tests/src/photo/photo.repository.ts#L5

### CODE REVIEW

File: lib/typeorm.providers.ts [hunk 1]
```diff
@@ -11,12 +11,13 @@ export function createTypeOrmProviders(
...
-  return (entities || []).map(entity => ({
+  return (entities || []).map((entity) => ({
     provide: getRepositoryToken(entity, connection),
     useFactory: (connection: Connection) => {
       if (
-        entity.prototype instanceof Repository ||
-        entity.prototype instanceof AbstractRepository
+        typeof Repository === 'function' &&
+        (entity.prototype instanceof Repository ||
+          entity.prototype instanceof AbstractRepository)
```
Comment (kamilmysliwiec) [inline lib/typeorm.providers.ts]: Btw what about custom repositories in 0.3.0. Are they going to be deprecated/removed?

Comment (matomesc) [reply lib/typeorm.providers.ts]: You can extend a repository with custom methods by calling `extend()` on a repository instance:

```ts
repository.extend({
  customMethod() {
     // ...
  }
});
```

https://github.com/typeorm/typeorm/blob/b177e230a6fcb11f1eb71d4d431d0297436b7f6f/src/repository/Repository.ts#L380

Comment (micalevisk) [reply lib/typeorm.providers.ts:20]: @kamilmysliwiec now (0.3.0) `@EntityRepository()` is deprecated. Here's how we should use instead: https://github.com/typeorm/typeorm/blob/master/docs/custom-repository.md

**Comment by kamilmysliwiec** [comment]
This PR is 2 years old so it may contain outdated code. Let me just close it and create a new one for migrating to the latest version of TypeORM

**Comment by derekcunningham-ab** [comment]
> This PR is 2 years old so it may contain outdated code. Let me just close it and create a new one for migrating to the latest version of TypeORM

@kamilmysliwiec Any luck with this? I'm starting a new project, and if NestJS is on the precipice of migrating, I'd rather commit to the latest version of TypeORM over using outdated approaches.

**Comment by kamilmysliwiec** [comment]
I've started 10 days ago and I noticed that TypeORM Readme & docs were still not even updated to the latest version so I figured I'll have to wait at least a few days more before I can jump onto this issue.

**Comment by lovetodream** [comment]
> > This PR is 2 years old so it may contain outdated code. Let me just close it and create a new one for migrating to the latest version of TypeORM
> 
> @kamilmysliwiec Any luck with this? I'm starting a new project, and if NestJS is on the precipice of migrating, I'd rather commit to the latest version of TypeORM over using outdated approaches.

@derekcunningham-ab I've created a [fork](https://github.com/lovetodream/typeorm) for myself, which seems to work fine as a temporary solution. Maybe you can use it until the official nestjs module is updated, if you don't want to wait.

**Comment by EdouardBougon** [comment]
It seems that the documentation is now updated

**Comment by bviala** [comment]
bump

**Comment by jmcdo29** [comment]
https://github.com/nestjs/typeorm/pull/1233

**Pull request closed without merging**